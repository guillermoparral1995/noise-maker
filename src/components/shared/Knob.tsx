import React, { BaseSyntheticEvent, useContext } from 'react';
import { stateContext } from '../../providers/StateProvider';
import type { ActionBuilder } from '../../types';

interface KnobProps {
  label: string;
  defaultValue: number;
  action: ActionBuilder<number>;
  min: number;
  max: number;
}

export const Knob = ({ label, defaultValue, action, min, max }: KnobProps) => {
  const { dispatch } = useContext(stateContext);
  const handleChange = (e: BaseSyntheticEvent) => {
    // range inputs return value as string even if params are specified as numbers
    dispatch(action(parseFloat(e.target.value)));
  };

  return (
    <>
      <input
        type="range"
        defaultValue={defaultValue}
        name={label}
        onChange={(e) => handleChange(e)}
        min={min}
        max={max}
        step={0.01}
      ></input>
      <label htmlFor={label}>{label}</label>
    </>
  );
};

export default Knob;
