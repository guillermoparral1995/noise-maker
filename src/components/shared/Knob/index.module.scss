.label {
    font-size: 12px;
    color: rgba(255, 255, 255, 0.6);
}