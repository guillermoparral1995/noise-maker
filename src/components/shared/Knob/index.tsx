import { round, throttle } from 'lodash';
import React, { KeyboardEvent, useCallback, useEffect, useState } from 'react';
import { WebAudioKnob } from 'webaudio-controls-react-typescript';
import { knobsValues } from '../../../constants/knobsValues';
import type { ActionTypes, Knobs } from '../../../types';
import styles from './index.module.scss';

interface KnobProps {
  id: Knobs;
  value: number;
  dispatch: React.Dispatch<ActionTypes>;
}

export const Knob = ({ id, value, dispatch }: KnobProps) => {
  const [isFocused, setIsFocused] = useState<boolean>(false);
  useEffect(() => {
    return () => throttledHandleChange.cancel();
  }, []);

  const handleChange = (e: number) => {
    dispatch(knobsValues[id].action(round(e as number, 2)));
  };

  const throttledHandleChange = useCallback(throttle(handleChange, 50), [
    dispatch,
  ]);

  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    if (isFocused) {
      if (event.code === 'ArrowUp') {
        dispatch(
          knobsValues[id].action(
            round(value + (knobsValues[id].step ?? 0.01), 2),
          ),
        );
      }

      if (event.code === 'ArrowDown') {
        dispatch(
          knobsValues[id].action(
            round(value - (knobsValues[id].step ?? 0.01), 2),
          ),
        );
      }
    }
  };

  const handleFocus = () => {
    setIsFocused(true);
  };

  const handleBlur = () => {
    setIsFocused(false);
  };

  return (
    <div
      onKeyDown={handleKeyDown}
      onFocus={handleFocus}
      onBlur={handleBlur}
      data-testid={id}
      tabIndex={-1}
    >
      <p className={styles.label}>{knobsValues[id].label}</p>
      <WebAudioKnob
        value={value}
        diameter={50}
        defvalue={knobsValues[id].default}
        min={knobsValues[id].min}
        max={knobsValues[id].max}
        step={knobsValues[id].step ?? 0.01}
        bodyColor="#f0f0f0"
        highlightColor="#ffffff"
        indicatorColor="mediumpurple"
        onKnobInput={throttledHandleChange}
        id={id}
      ></WebAudioKnob>
      <p className={styles.label}>{value}</p>
    </div>
  );
};

export default Knob;
