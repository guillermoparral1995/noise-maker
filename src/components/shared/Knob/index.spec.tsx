import {
  act,
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import React from 'react';
import Knob from '.';
import '@testing-library/jest-dom';
import { Actions, Knobs } from '@types';

jest.mock('@hooks/useBreakpoints', () => ({
  useBreakpoints: () => ({
    isMobile: false,
  }),
}));

describe('Knob', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render with knob value', async () => {
    const { findByTestId } = render(
      <Knob id={Knobs.VOLUME} value={0.5} dispatch={jest.fn()}></Knob>,
    );

    const knob = await findByTestId(Knobs.VOLUME);
    expect(screen.getByText('0.5')).toBeVisible();
    expect(knob).toMatchSnapshot();
  });

  /**
   * Testing this component on click events is extremely hard due to
   * internal PrimeReact implementation.
   * Keyboard interaction was added for accessibility and easier testing
   */
  it('should call dispatch when changing value - arrow up', async () => {
    const mockDispatch = jest.fn();
    render(<Knob id={Knobs.VOLUME} value={0.5} dispatch={mockDispatch}></Knob>);

    const knob = screen.getByTestId(Knobs.VOLUME);
    act(() => {
      knob.focus();
    });
    fireEvent.keyDown(knob, { key: 'ArrowUp', code: 'ArrowUp' });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: 0.51,
      type: Actions.UPDATE_VOLUME,
    });
  });

  it('should call dispatch when changing value - arrow down', async () => {
    const mockDispatch = jest.fn();
    render(<Knob id={Knobs.VOLUME} value={0.5} dispatch={mockDispatch}></Knob>);

    const knob = screen.getByTestId(Knobs.VOLUME);
    act(() => {
      knob.focus();
    });
    fireEvent.keyDown(knob, { key: 'ArrowDown', code: 'ArrowDown' });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: 0.49,
      type: Actions.UPDATE_VOLUME,
    });
  });
});
