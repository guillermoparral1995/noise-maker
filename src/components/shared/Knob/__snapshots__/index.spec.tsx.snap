// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Knob should render with knob value 1`] = `
<div
  data-testid="VOLUME"
  tabindex="-1"
>
  <p
    class="label"
  >
    Volume
  </p>
  <webaudio-knob
    bodycolor="#f0f0f0"
    colors="mediumpurple;#f0f0f0;#ffffff;"
    defvalue="0.5"
    diameter="50"
    highlightcolor="#ffffff"
    id="VOLUME"
    indicatorcolor="mediumpurple"
    max="1"
    min="0"
    step="0.01"
    style="height: 50px;"
    value="0.5"
  />
  <p
    class="label"
  >
    0.5
  </p>
</div>
`;
